### DISCUSSION TITLE
git fetch on tag builds does not fetch the tag that initiated the build

### DISCUSSION THREAD

**Discussion by asbjornu** [community/community#25159] [Actions]
<p>I’m seeing some really weird behavior when subscribing to the push: tags event:</p>
<pre><code>on:
  push:
    tags:
      - '*'
</code></pre>
<p>When a build is initiated from the creation of tag, a step doing git fetch --prune --unshallow --tags does not fetch the tag that initiated the build. <a href="https://github.com/SwedbankPay/kramdown-plantuml/runs/550358440#step:3:11" rel="nofollow noopener">This build</a> was initiated from creating tag 1.0.3 as can be seen by the refs/tags/1.0.3 value of github.ref. But as the output from git fetch shows, tag 1.0.3 isn’t fetched:</p>
<pre><code>From https://github.com/SwedbankPay/kramdown-plantuml
* [new branch] develop -&gt; origin/develop
* [new branch] master -&gt; origin/master
* [new tag] 1.0.0 -&gt; 1.0.0
* [new tag] 1.0.1 -&gt; 1.0.1
* [new tag] 1.0.2 -&gt; 1.0.2
</code></pre>
<p>Not having access to the tag makes <a href="https://github.com/marketplace/actions/use-actions" rel="nofollow noopener">GitVersion</a> not work properly, generating the version number 1.1.0-tags-1-0-3.1+15 instead of 1.0.3.</p>
<p>Can someone please explain why the tag that initiated the build is not available to the build itself?</p>
<p>PS: I sincerely hate the forum software you’ve chosen here and the fact that something so specifically targeted at developers makes it impossible to use Markdown or HTML to format messages. I enjoy writing GitHub issues 178 times more than writing into this godforsaken “rich text” field. Please, burn this to the ground. Thank you.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/asbjornu">@asbjornu</a>,</p>
<p>The default fetch depth is ‘1’ for action ‘actions/checkout@v2’, it will checkout the latest tag(1.0.3). Hence, when you run command ‘<span>git fetch --prune --unshallow --tags</span>’, it’ll get the non-existing tags which ends with tag ‘1.0.2’.  I confirmed the appearance is same on my local git env.</p>
<p>For GitVersion use, please check ticket: <a href="https://github.community/t/checkout-action-does-not-create-local-master-and-has-no-options-for-tags-breaking-gitversion/16239">https://github.community/t5/GitHub-Actions/Checkout-Action-does-not-create-local-master-and-has-no-options/m-p/31575</a></p>
<p>Hope it helps!</p>

**Comment by asbjornu** [comment]
<p>Thanks for the information, <a class="mention" href="/weide-zhou">@weide-zhou</a>. I was not aware of fetch-depth: 0. <a href="https://github.com/SwedbankPay/kramdown-plantuml/commit/05723113a9d0822d9b92dc4fbbb1a80f3ecd3925" rel="nofollow noopener">However, setting that now leads to GitHub Actions not being triggered at all</a>, for whatever reason. A temporary glitch in the matrix? Also, from the issue you linked, it seems that it should be fixed in v2 of the checkout action:</p>
<blockquote>
<p>…there will be a 2.0 version coming that will make sure you are not in a detached head state.</p>
</blockquote>
<p>What I don’t understand, still, is how git fetch on the same repository that the tag 1.0.3 was pushed to, for builds triggered by the creation of 1.0.3, can’t fetch tag 1.0.3.</p>

**Comment by asbjornu** [comment, marked as answer]
<p>I ended up just using the git tag directly and disable GitVersion when the build is triggered from a tag with if: startsWith(github.ref, ‘refs/tags/’) != true. I perform the stable/unstable versioning logic with actions/github-script as follows:</p>
<pre><code>const gemVersion = (function() {
  const ref = '${{ github.ref }}';
  const tagPrefix = 'refs/tags/';

  if (ref.startsWith(tagPrefix)) {
    // If a tag ref is being built, just return the tag verbatim
    return ref.substring(tagPrefix.length);
  }

  const escapedBranchName = '${{ steps.gitversion.outputs.escapedBranchName }}';
  const commitsSinceVersionSource = '${{ steps.gitversion.outputs.commitsSinceVersionSourcePadded }}';
  const preReleaseLabel = escapedBranchName + '.' + commitsSinceVersionSource;
  const majorMinorPatch = '${{ steps.gitversion.outputs.majorMinorPatch }}';
  return majorMinorPatch + preReleaseLabel;
})();

core.setOutput('version', gemVersion);
</code></pre>
<p>It would be much easier if I could just do git fetch --tags in the build to have the tag being built locally, but as that seems impossible, this is at least a suitable workaround.</p>
